src/main/scala/com/microsoft/hyperspace/index/rankers/JoinIndexRanker.scala

```diff
@@ -16,40 +16,78 @@
 
 /**
- * Ranker class for Join rule indexes
+ * Ranker class for Join rule indexes.
  */
 object JoinIndexRanker {
 
   /**
    * Rearranges all available index options according to their cost. The first one is the best
    * with minimum cost.
    *
-   * Pick the one with least amount of shuffling and most number of buckets. When two indices have
-   * same number of buckets, there is zero shuffling. If number of buckets differ, one of the
-   * indices gets reshuffled into the number of buckets equal to the other. Secondly, more the
-   * number of buckets, better the parallelism achieved during join in general, assuming there is
-   * no resource constraint.
+   * If hybridScanEnabled is false, pick the one with least amount of shuffling and most number
+   * of buckets. When two indices have same number of buckets, there is zero shuffling.
+   * If number of buckets differ, one of the indices gets reshuffled into the number of buckets
+   * equal to the other.
+   * Secondly, more the number of buckets, better the parallelism achieved during join in general
+   * assuming there is no resource constraint.
+   *
+   * If hybridScanEnabled is true, rank algorithm follows the algorithm above, but we prioritize
+   * the index with larger common source data, so that we could minimize the amount of data
+   * for on-the-fly shuffle or merge.
    *
-   * @param indexPairs index pairs for left and right side of the join. All index pairs are
+   * @param spark SparkSession.
+   * @param leftChild Logical relation of left child of the join.
+   * @param rightChild Logical relation of right child of the join.
+   * @param indexPairs Index pairs for left and right side of the join. All index pairs are
    *                   compatible with each other.
-   * @return rearranged index pairs according to their ranking. The first is the best.
+   * @return Rearranged index pairs according to their ranking. The first is the best.
    */
   def rank(
+      spark: SparkSession,
+      leftChild: LogicalPlan,
+      rightChild: LogicalPlan,
       indexPairs: Seq[(IndexLogEntry, IndexLogEntry)]): Seq[(IndexLogEntry, IndexLogEntry)] = {
-
+    val hybridScanEnabled = HyperspaceConf.hybridScanEnabled(spark)
     indexPairs.sortWith {
-      case ((left1, left2), (right1, right2)) =>
-        if (left1.numBuckets == left2.numBuckets && right1.numBuckets == right2.numBuckets) {
-          left1.numBuckets > right1.numBuckets
-        } else if (left1.numBuckets == left2.numBuckets) {
+      case ((left1, right1), (left2, right2)) =>
+        // These common bytes were calculated and tagged in getCandidateIndexes.
+        // The value is the summation of common source files of the given plan and each index.
+        lazy val leftCommonBytes = left1
+          .getTagValue(leftChild, IndexLogEntryTags.COMMON_BYTES)
+          .getOrElse(0L) + right1
+          .getTagValue(rightChild, IndexLogEntryTags.COMMON_BYTES)
+          .getOrElse(0L)
```

Comment:
can we create a local def to simplify this? `lazy val leftCommonBytes = getCommonBytes(left1, leftChild) + getCommonBytes(right1, rightChild)`, for example